pkg/hierarchy/manager_test.go

```diff
@@ -0,0 +1,232 @@
+				m.AddClusterQueue(newCq("queue1"))
+				m.AddClusterQueue(newCq("queue2"))
+				m.UpdateClusterQueueEdge("queue1", "cohort-a")
+				m.UpdateClusterQueueEdge("queue2", "cohort-b")
+
+				m.UpdateClusterQueueEdge("queue1", "cohort-b")
+				m.UpdateClusterQueueEdge("queue2", "cohort-c")
+				m.UpdateClusterQueueEdge("queue1", "cohort-b")
+				m.UpdateClusterQueueEdge("queue2", "cohort-c")
+				m.UpdateClusterQueueEdge("queue2", "cohort-c")
+				m.UpdateClusterQueueEdge("queue1", "cohort-b")
+			},
+			wantCqs:     sets.New("queue1", "queue2"),
+			wantCohorts: sets.New("cohort-b", "cohort-c"),
+			wantCqEdge: []cqEdge{
+				{"queue1", "cohort-b"},
+				{"queue2", "cohort-c"},
+			},
+		},
+		"delete cohorts": {
+			operations: func(m M) {
+				m.AddClusterQueue(newCq("queue1"))
+				m.AddClusterQueue(newCq("queue2"))
+
+				m.UpdateClusterQueueEdge("queue1", "cohort-a")
+				m.UpdateClusterQueueEdge("queue2", "cohort-b")
+
+				// Delete cohort-a by deleting edge.
+				// Delete cohort-b by deleting cq.
+				m.UpdateClusterQueueEdge("queue1", "")
+				m.DeleteClusterQueue("queue2")
+			},
+			wantCqs: sets.New("queue1"),
+		},
+		"deletion idempotent": {
+			operations: func(m M) {
+				m.DeleteClusterQueue("doesnt-exist")
+
+				m.AddClusterQueue(newCq("deleted-once"))
+				m.AddClusterQueue(newCq("deleted-twice"))
+
+				m.DeleteClusterQueue("doesnt-exist2")
+				m.DeleteClusterQueue("deleted-once")
+				m.DeleteClusterQueue("deleted-twice")
+				m.DeleteClusterQueue("deleted-twice")
+			},
+			wantCqs: sets.New[string](),
+		},
+		"cohort remains as long as one of its children remains": {
+			operations: func(m M) {
+				m.AddClusterQueue(newCq("queue1"))
+				m.AddClusterQueue(newCq("queue2"))
+				m.AddClusterQueue(newCq("queue3"))
+				m.UpdateClusterQueueEdge("queue1", "cohort")
+				m.UpdateClusterQueueEdge("queue2", "cohort")
+				m.UpdateClusterQueueEdge("queue3", "to-be-deleted-cohort")
+
+				m.DeleteClusterQueue("queue2")
+				m.DeleteClusterQueue("queue3")
+			},
+			wantCqs:     sets.New("queue1"),
+			wantCohorts: sets.New("cohort"),
+			wantCqEdge:  []cqEdge{{"queue1", "cohort"}},
+		},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			mgr := NewManager[*testClusterQueue, *testCohort](cohortFactory)
+			tc.operations(mgr)
+			t.Run("clusterqueues match", func(t *testing.T) {
+				gotCqs := sets.New[string]()
+				for _, cq := range mgr.ClusterQueues {
+					gotCqs.Insert(cq.GetName())
+				}
+				if diff := cmp.Diff(tc.wantCqs, gotCqs); diff != "" {
+					t.Fatalf("Unexpected cqs -want +got %s", diff)
+				}
+			})
+			t.Run("cohorts match", func(t *testing.T) {
+				gotCohorts := sets.New[string]()
+				for _, cohort := range mgr.Cohorts {
+					gotCohorts.Insert(cohort.GetName())
+				}
+				if diff := cmp.Diff(tc.wantCohorts, gotCohorts); diff != "" {
+					t.Fatalf("Unexpected cohorts -want +got %s", diff)
+				}
+			})
+			t.Run("verify cq<->cohort edge count", func(t *testing.T) {
```

Comment:
I have a feeling these two tests "verify cq<->cohort edge count" and "verify cq<->cohort edges" are quite verbose. Would it be simpler to create a set of pairs edges by iterating over the graph, then use cmp.Diff to compare it with `wantCqEdge`?

Reply:
Yes, I quite like this suggestion. It is much simpler now :)

 Thanks!

Reply:
lgtm, thanks